### ISSUE TITLE
kube-proxy stuck in ContainerCreating

### ISSUE THREAD

**Issue by jackfrancis** [Azure/aks-engine#679]
We're seeing this regularly across our E2E tests:

```
kube-system   kube-proxy-29jv2                                1/1     Running             0          32m     10.239.255.241   k8s-master-35441517-2   <none>           <none>
kube-system   kube-proxy-5zf7s                                1/1     Running             0          3m27s   10.239.2.35      k8s-agent2-35441517-1   <none>           <none>
kube-system   kube-proxy-996r5                                1/1     Running             0          11m     10.239.1.4       k8s-agent1-35441517-1   <none>           <none>
kube-system   kube-proxy-bx6gc                                0/1     ContainerCreating   0          46m     10.239.255.239   k8s-master-35441517-0   <none>           <none>
kube-system   kube-proxy-xsmrn                                1/1     Running             0          40m     10.239.255.240   k8s-master-35441517-1   <none>           <none>
```

See a single kube-proxy not able to be fully reconciled on a single node.

Pod logs aren't interesting:

```
2019/03/07 18:40:59 $ k logs kube-proxy-bx6gc -c kube-proxy -n kube-system
2019/03/07 18:40:59 #### $ k logs kube-proxy-bx6gc -c kube-proxy -n kube-system completed in 376.260953ms
2019/03/07 18:40:59 
Error from server (BadRequest): container "kube-proxy" in pod "kube-proxy-bx6gc" is waiting to start: ContainerCreating
```

**Comment by jackfrancis** [comment]
@feiskyer @andyzhangx @ritazh have you seen the above symptoms in any upstream-originating k8s e2e tests?

**Comment by feiskyer** [comment]
@jackfrancis Could you run kubectl describe for the pod? There may be something wrong when creating the pod sandbox (usually because of network plugins issues)

**Comment by jackfrancis** [comment]
@feiskyer will do, see:

https://github.com/Azure/aks-engine/pull/691

Do pod sandbox creation errors due to network plugin issues reconcile themselves? What would be the expected remediation steps if we see that?

**Comment by feiskyer** [comment]
It depends the errors. sometimes delete pod and let the controller recreate a new one may help.

**Comment by jackfrancis** [comment]
@feiskyer here's a repro w/ kubectl describe pod:

```
2019/03/18 19:09:57 #### $ k describe pod kube-proxy-fhnc9 -n kube-system completed in 394.489276ms
2019/03/18 19:09:57 
Name:               kube-proxy-fhnc9
Namespace:          kube-system
Priority:           2000001000
PriorityClassName:  system-node-critical
Node:               k8s-master-35355342-0/10.255.255.5
Start Time:         Mon, 18 Mar 2019 17:32:25 +0000
Labels:             component=kube-proxy
                    controller-revision-hash=7c849bc58f
                    pod-template-generation=1
                    tier=node
Annotations:        <none>
Status:             Pending
IP:                 10.255.255.5
Controlled By:      DaemonSet/kube-proxy
Containers:
  kube-proxy:
    Container ID:  
    Image:         k8s.gcr.io/hyperkube-amd64:v1.12.6
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      /hyperkube
      proxy
      --kubeconfig=/var/lib/kubelet/kubeconfig
      --cluster-cidr=10.240.0.0/12
      --feature-gates=ExperimentalCriticalPodAnnotation=true
      --proxy-mode=iptables
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:        100m
    Environment:  <none>
    Mounts:
      /etc/kubernetes from etc-kubernetes (ro)
      /etc/ssl/certs from ssl-certs-host (ro)
      /lib/modules/ from kernelmodules (ro)
      /run/xtables.lock from iptableslock (rw)
      /var/lib/kubelet/kubeconfig from kubeconfig (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cbf87 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  ssl-certs-host:
    Type:          HostPath (bare host directory volume)
    Path:          /usr/share/ca-certificates
    HostPathType:  
  kubeconfig:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/kubeconfig
    HostPathType:  
  etc-kubernetes:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/kubernetes
    HostPathType:  
  iptableslock:
    Type:          HostPath (bare host directory volume)
    Path:          /run/xtables.lock
    HostPathType:  
  kernelmodules:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/modules/
    HostPathType:  
  default-token-cbf87:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-cbf87
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  beta.kubernetes.io/os=linux
Tolerations:     :NoExecute
                 :NoSchedule
                 node-role.kubernetes.io/master=true:NoSchedule
                 node.kubernetes.io/disk-pressure:NoSchedule
                 node.kubernetes.io/memory-pressure:NoSchedule
                 node.kubernetes.io/network-unavailable:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute
                 node.kubernetes.io/unreachable:NoExecute
                 node.kubernetes.io/unschedulable:NoSchedule
Events:
  Type     Reason      Age                    From                            Message
  ----     ------      ----                   ----                            -------
  Normal   Scheduled   97m                    default-scheduler               Successfully assigned kube-system/kube-proxy-fhnc9 to k8s-master-35355342-0
  Normal   Pulled      97m                    kubelet, k8s-master-35355342-0  Container image "k8s.gcr.io/hyperkube-amd64:v1.12.6" already present on machine
  Normal   Created     97m                    kubelet, k8s-master-35355342-0  Created container
  Normal   Started     97m                    kubelet, k8s-master-35355342-0  Started container
  Normal   Pulled      33m                    kubelet, k8s-master-35355342-0  Container image "k8s.gcr.io/hyperkube-amd64:v1.12.6" already present on machine
  Normal   Created     33m                    kubelet, k8s-master-35355342-0  Created container
  Normal   Started     33m                    kubelet, k8s-master-35355342-0  Started container
  Warning  FailedSync  3m10s (x139 over 33m)  kubelet, k8s-master-35355342-0  error determining status: rpc error: code = Unknown desc = unable to inspect docker image "sha256:1a40bf769d4b5298bada01e8532b93a0ea6d9b82241defa1bc58666e48fb5827" while inspecting docker container "b2bdcdfcd5792980c7899df7e404305eb091c447a6a14da09e3bb4a4324ef3fd": no such image: "sha256:1a40bf769d4b5298bada01e8532b93a0ea6d9b82241defa1bc58666e48fb5827"
```

**Comment by jackfrancis** [comment]
@cpuguy83 does the `FailedSync` message at the bottom of the above `kubectl describe pod` output suggest the root cause of this symptom is the container runtime? If so, is there an appropriate nanny-like vector we could use to identify such symptoms, and to automatically take remediation action?

**Comment by cpuguy83** [comment]
@jackfrancis It's hard to say. Just from the message it would seem that dockershim thinks an image should be there which is not actually there.

**Comment by cpuguy83** [comment]
I we have debug logs from the engine (hopefully without tons of noise unrelated to the test) it should at least be possible to tell what API calls have been made.

**Comment by jackfrancis** [comment]
Happy to add add'l debug log gathering to our E2E tests (that's where we're seeing these intermittent symptoms) and continue to populate this issue :)

What should we grab next time before the E2E runner destroys the cluster?

**Comment by cpuguy83** [comment]
Just enable debug logging (`-D` flag on dockerd) and grab whatever file this is set to log to... though I suppose it may be in journald.

**Comment by feiskyer** [comment]
The error message indicates kubelet doesn't find the docker image used by a container. This is actually weird because a used image couldn't be deleted anyway.

**Comment by jackfrancis** [comment]
I have a lead...

We deliver kube-proxy via addons (via the filesystem):

```
$ ls -la /etc/kubernetes/addons/ | grep 'proxy'
-rw-r--r-- 1 root root 2046 Mar 19 20:50 kube-proxy-daemonset.yaml
```

In my repro case, the cluster has been upgraded from 1.11 to 1.12:

```
$ kubectl get nodes
NAME                    STATUS   ROLES    AGE    VERSION
k8s-agent1-81894314-1   Ready    agent    67m    v1.12.6
k8s-agent2-81894314-1   Ready    agent    59m    v1.12.6
k8s-master-81894314-0   Ready    master   81m    v1.12.6
k8s-master-81894314-1   Ready    master   76m    v1.12.6
k8s-master-81894314-2   Ready    master   144m   v1.12.6
```

But the kube-proxy spec on the failed instance suggests the daemonset hasn't been cleaned up:

```
$ kubectl describe pod kube-proxy-rd42k -n kube-system | grep hyperkube
    Image:         k8s.gcr.io/hyperkube-amd64:v1.11.8
      /hyperkube
```

To verify, the daemonset spec in `/etc/kubernetes/addons/` is correct after upgrade:

```
$ grep hyperkube /etc/kubernetes/addons/kube-proxy-daemonset.yaml 
        - /hyperkube
        image: k8s.gcr.io/hyperkube-amd64:v1.12.6
```

**Comment by jackfrancis** [comment]
Simply deleting the pod doesn't work:

```
$ kubectl delete pod kube-proxy-rd42k -n kube-system
pod "kube-proxy-rd42k" deleted

```

(Hangs...)

**Comment by jackfrancis** [comment]
This solved the problem:

```
$ kubectl delete pod kube-proxy-rd42k -n kube-system --grace-period=0 --force
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
pod "kube-proxy-rd42k" force deleted
```

tl;dr if you have an orphaned pod from a <daemonset|deployment|etc> that no longer exists, force delete may be the trick

**Comment by jackfrancis** [comment]
I think I may have identified what's going on here.

The kube-proxy daemonset that is scheduled to a master node isn't reconciled as we expect during upgrade, which means that the daemonset spec is asking for the prior version of k8s (kube-proxy image is delivered via hyperkube image), for example if we upgrade from 1.12.8 to 1.13.9 the master node(s) come online with a kube-proxy running @ version 1.12.8. In our CSE script we immediately clean up all hyperkube images at the end of cluster bootstrapping. As part of our validation of a master node's CSE script, we bring the kubelet online. So what's happening:

1. kubelet comes online and checks in as schedulable, and images begin to get scheduled onto the upgraded master node
2. kube-proxy daemonset pod gets scheduled with a spec describing an old version of k8s
3. If that old version of k8s is actually pre-baked into the VHD that the node vm was built with, then the pod scheduling execution may enter a race condition with the CSE-induced container image cleanup implementation; for example, the daemonset may be scheduled prior to image deletion, and so the container runtime will say "I already have this image!", and then a few ticks later CSE deletes the image, while the kubelet is still in the process of scaffolding a runtime from that pre-existing image reference (an image reference that now longer exists).
